packages/react/src/jsx.tsx

```diff
@@ -49,24 +48,35 @@ export function jsx(
   }
 }
 
-export type Extend<T> = { [P in keyof T]?: _Extend<T[P]> }
+export type Literal<T> = {
+  __literal: T
+}
+export type LiteralArray<T> = Array<T> & {
+  __literalArray: T[]
+}
 
-// eslint-disable-next-line @typescript-eslint/ban-types
-export type _Extend<T> = T extends Function
-  ? T | ServerEventDataState | ActionDataState
-  : T extends Array<infer U>
-    ? _DeepPartialArray<U>
-    : T extends object
-      ? Extend<T>
-      : T | ReferencedDataState | StateDataState
+export type WithServerProps<T> = { [P in keyof T]: _Extend<T[P]> }
+interface _ExtendArray<T> extends Array<_Extend<T>> {}
 
-export interface _DeepPartialArray<T> extends Array<_Extend<T>> {}
+export type _Extend<T> = T extends StateDataState
```

Comment:
thanks for explaining this to me earlier. we should probably write a big comment at some point that explains for newcomers to the code

Reply:
we don't really have any documentation mechanism yet in place, ideally we can come up with something that allows us to write documentation inline alongside relevant code